### PR TITLE
feat(PeriphDrivers): Add clock selection feature to WDT

### PR DESCRIPTION
**Pull request by hfakkiz**
### Description

WDT peripheral of MAX32655, MAX32662, MAX32670, MAX32672, MAX32675, MAX32680, MAX32690, MAX78000 and MAX78002 supports selecting different clock sources which are given in user guide. 
Added a new function to set clksel register on revB WDT driver.
Added a new function to select supported clock sources for these part numbers.

### Checklist Before Requesting Review

- [x] PR Title follows correct guidelines.
- [x] Description of changes and all other relevant information.
- [ ] (Optional) Link any related GitHub issues [using a keyword](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)
- [ ] (Optional) Provide info on any relevant functional testing/validation.  For API changes or significant features, this is not optional.

### CODE REVIEW

File: Libraries/PeriphDrivers/Source/WDT/wdt_ai87.c
Comment (Jake-Carter) [inline Libraries/PeriphDrivers/Source/WDT/wdt_ai87.c:166]: For micros where all WDT instances have the same available clock sources you can reduce this

```C
int MXC_WDT_SetClockSource(mxc_wdt_regs_t *wdt, mxc_wdt_clock_t clock_source)
{
    const uint8_t clock_source_num = 8;
    uint8_t idx = 0;
    uint8_t instance = 0;
    mxc_wdt_clock_t clock_sources[2][8] = {
        { MXC_WDT_PCLK, MXC_WDT_IPO_CLK, MXC_WDT_IBRO_CLK, MXC_WDT_INRO_CLK, MXC_WDT_ERTCO_CLK,
          MXC_WDT_EXT_CLK, MXC_WDT_ERFO_CLK, 0xFF },
        { MXC_WDT_PCLK, MXC_WDT_IPO_CLK, MXC_WDT_IBRO_CLK, MXC_WDT_INRO_CLK, MXC_WDT_ERTCO_CLK,
          MXC_WDT_EXT_CLK, MXC_WDT_ERFO_CLK, 0xFF }
    };

    if (wdt == MXC_WDT0) {
        instance = 0;
    } else if (wdt == MXC_WDT1) {
        instance = 1;
    } else {
        return E_BAD_PARAM;
    }

    for (idx = 0; idx < clock_source_num; idx++) {
        if (clock_sources[instance][idx] == clock_source) {
            break;
        }
    }

    if (idx == clock_source_num) {
        return E_BAD_PARAM;
    }

    MXC_WDT_RevB_SetClockSource((mxc_wdt_revb_regs_t *)wdt, idx);

    return E_NO_ERROR;
}
```

to this.  It will help reduce unnecessary code size.  

I'd also suggest adding a note to the `mxc_wdt_clock_t` definition that the ordering of the enum should match the register field exactly.

```C
int MXC_WDT_SetClockSource(mxc_wdt_regs_t *wdt, mxc_wdt_clock_t clock_source)
{
    MXC_WDT_RevB_SetClockSource((mxc_wdt_revb_regs_t *)wdt, (int)clock_source);

    return E_NO_ERROR;
}
```

File: Libraries/PeriphDrivers/Source/WDT/wdt_me12.c
Comment (Jake-Carter) [inline Libraries/PeriphDrivers/Source/WDT/wdt_me12.c:171]: You could also apply the same reduction here, since there is only a single WDT instance

File: Libraries/PeriphDrivers/Source/WDT/wdt_me17.c
Comment (Jake-Carter) [inline Libraries/PeriphDrivers/Source/WDT/wdt_me17.c:174]: Throw a build error here instead with `#error ...`.  This will help us catch cases where this file is reused in the future.

File: Libraries/PeriphDrivers/Source/WDT/wdt_me18.c
Comment (Jake-Carter) [inline Libraries/PeriphDrivers/Source/WDT/wdt_me18.c:159]: Why's `clock_source_num` always 8?

File: Libraries/PeriphDrivers/Include/MAX32655/wdt.h
Comment (Jake-Carter) [inline Libraries/PeriphDrivers/Include/MAX32655/wdt.h:131]: It'd be ideal if we didn't need to create this, and could just use `mxc_sys_system_clock_t` as the second parameter to the function.

```C
/**
 * @brief       Sets clock source.
 * @param       wdt             Pointer to watchdog registers.
 * @param       clock_source    Clock source.
 */
int MXC_WDT_SetClockSource(mxc_wdt_regs_t *wdt, mxc_system_clock_t clock_source);
```

The default PCLK option is problematic though...  no action needed on this unless you have any ideas.  The less enums we have to maintain that are specific to each micro the better

**Review by Jake-Carter — CHANGES_REQUESTED** [review]
Thanks @hfakkiz, looks good.  A couple minor comments below

File: Libraries/PeriphDrivers/Source/WDT/wdt_me18.c
Comment (hfakkiz) [reply Libraries/PeriphDrivers/Source/WDT/wdt_me18.c:159]: WDT supports 8 different clock sources but mostly not using all of 8 options. I set `clock_source_num` parameter and `clock_sources` array size to 8 because upcoming revisions can use other options which are not used now. To reduce code size, I can change them to number of supported clock source.

File: Libraries/PeriphDrivers/Source/WDT/wdt_me17.c
Comment (hfakkiz) [reply Libraries/PeriphDrivers/Source/WDT/wdt_me17.c]: Updated code for ME15 and ME17.

File: Libraries/PeriphDrivers/Source/WDT/wdt_ai87.c
Comment (hfakkiz) [reply Libraries/PeriphDrivers/Source/WDT/wdt_ai87.c]: Updated code for ME12, ME15, ME21 and AI87.

File: Libraries/PeriphDrivers/Include/MAX32655/wdt.h
Comment (hfakkiz) [reply Libraries/PeriphDrivers/Include/MAX32655/wdt.h:131]: As you said PCLK option is a problem for this solution. Also, using `mxc_wdt_clock_t` enum helps to use only supported clock sources of WDT. If it's not a problem, I prefer to use `mxc_wdt_clock_t` enum solution.

File: Libraries/PeriphDrivers/Source/WDT/wdt_me18.c
Comment (Jake-Carter) [inline Libraries/PeriphDrivers/Source/WDT/wdt_me18.c:162]: Initialize with `clock_source_num` instead of `8`

```C
mxc_wdt_clock_t clock_sources[2][clock_source_num] = ...
```

**Review by Jake-Carter — CHANGES_REQUESTED** [review]
Thank you @hfakkiz, one last change request based on your latest comments

File: Libraries/PeriphDrivers/Source/WDT/wdt_revb.c
Comment (hfakkiz) [reply Libraries/PeriphDrivers/Source/WDT/wdt_revb.c:159]: Done

File: Libraries/PeriphDrivers/Source/WDT/wdt_me18.c
Comment (Jake-Carter) [reply Libraries/PeriphDrivers/Source/WDT/wdt_me18.c:162]: Hmm strange...  should work as long as it's `const`.  What's the error?

Comment (hfakkiz) [reply Libraries/PeriphDrivers/Source/WDT/wdt_me18.c:162]: It is "error: variable-sized object may not be initialized".

image

**Comment by Jake-Carter** [comment]
Thanks @hfakkiz, looks good to go.  We'll get a second review then merge.

File: Libraries/PeriphDrivers/Source/WDT/wdt_me18.c
Comment (ozersa) [inline Libraries/PeriphDrivers/Source/WDT/wdt_me18.c:181]: It seems it will be nice if we have some utilities functions place, for this case, if we have a function like below in some place it will simplify me17/me18... source codes.
`
int mxc_util_get_index(int item, int *arr, int num_of_item)
{
    int i;

    for (i = (num_of_item-1); i >= 0; i--) {
        if (arr[i] == item) {
            break;
        }
    }
    
    return i; // on success return index, on fail return -1;
}
`

File: Libraries/PeriphDrivers/Source/WDT/wdt_revb.c
Comment (ozersa) [inline Libraries/PeriphDrivers/Source/WDT/wdt_revb.c:162]: Please use parenthesis

File: Libraries/PeriphDrivers/Source/WDT/wdt_me18.c
Comment (ozersa) [inline Libraries/PeriphDrivers/Source/WDT/wdt_me18.c:179]: It seems it will be nice if we have some utilities functions place, for this case, if we have a function like below in some place it will simplify me17/me18... source codes.
`
int mxc_util_get_index(int item, int *arr, int num_of_item)
{
    int i;

    for (i = (num_of_item-1); i >= 0; i--) {
        if (arr[i] == item) {
            break;
        }
    }
    
    return i; // on success return index, on fail return -1;
}
`

Comment (hfakkiz) [reply Libraries/PeriphDrivers/Source/WDT/wdt_me18.c:179]: For ME17 and ME18, `clock_sources` arrays have 2 or 3 elements from the beginning of arrays. For loop solution would find index in maximum 3 iterations. Using a search function like `lfind(...)` instead of this for loop will make it much more complicated.

File: Libraries/PeriphDrivers/Source/WDT/wdt_revb.c
Comment (hfakkiz) [reply Libraries/PeriphDrivers/Source/WDT/wdt_revb.c]: Done.

**Review by Jacob-Scheiffler — APPROVED** [review]

**Review by ozersa — APPROVED** [review]

**Pull request merged**
